Formal2024/Workshops/workshop3.lean

```diff
@@ -0,0 +1,210 @@
+  constructor
+  · intro h' heq
+    apply h'
+    rw [heq]
+  · intro h' hle
+    apply h'
+    exact le_antisymm h hle
+
+
+-- Lean knows that `↔` is reflexive, symmetric, and transitive, so you can use the `rw` tactic with it.
+#check abs_lt
+
+example (x : ℝ) : |x + 3| < 5 → -8 < x ∧ x < 2 := by
+  rw [abs_lt]
+  intro h
+  constructor <;> linarith
+-- The `<;>` tactic combinator attempts to close every open goal with the given tactic
+
+
+/-!
+  # Exercises
+-/
+
+-- 1. Try to prove this lemma in at least two different ways, e.g. using `constructor`, `⟨⟩`, `h.left` and `h.right`, etc. [MIL 3.4]
+
+example {x y : ℝ} : x ≤ y ∧ ¬y ≤ x ↔ x ≤ y ∧ x ≠ y := by
+  sorry
+
+
+-- 2. Prove the following lemma. [MIL 3.4]
+
+example : ¬Monotone fun x : ℝ ↦ -x := by
+  rw [Monotone] -- This unfolds the definition of `Monotone` in the goal.
+  push_neg
+  sorry
+
+
+/-!
+  # Disjunction
+-/
+
+-- The symbol for disjunction is `∨`, which you can write as `\or`.
+
+-- To prove a disjunction, it is enough to prove one of the two sides.
+-- You can use the `left` or `right` to choose which side you want to prove.
+
+example {x y : ℝ} (h : y > x ^ 2) : y > 0 ∨ y < -1 := by
+  left
+  linarith [pow_two_nonneg x] -- You can add extra hypotheses to `linarith` with square brackets.
+  -- In this case we have added `x^2 ≥ 0`.
+
+-- The proof term version of this is `Or.inl` and `Or.inr`.
+example {x y : ℝ} (h : y > x ^ 2) : y > 0 ∨ y < -1 :=
+  Or.inl (by linarith [pow_two_nonneg x])
+
+
+-- To use a disjunction, you can use `rcases` or `rintro`, as with existential quantifiers and conjunctions.
+-- The notation is slightly different: you use `|` instead of `⟨⟩`.
+-- When we use a disjunction, we have to prove the goal for each side of the disjunction.
+example {x y : ℝ} : x < |y| → x < y ∨ x < -y := by
+  rcases le_or_gt 0 y with h | h -- `le_or_gt` is a lemma that says `0 ≤ y ∨ y < 0`.
+  · rw [abs_of_nonneg h]
+    intro h; exact Or.inl h
+  . rw [abs_of_neg h]
+    intro h; exact Or.inr h
+
+-- You can nest patterns in `rcases` and `rintro`, even mixing `|` and `⟨⟩`.
+example {m n k : ℕ} (h : m ∣ n ∨ m ∣ k) : m ∣ n * k := by
+  rcases h with ⟨a, rfl⟩ | ⟨b, rfl⟩
+  -- Recall that using `rfl` in a pattern will rewrite the equation it replaces everywhere.
+  -- Here the first `rfl` will replace `n` with `m * a`, and second will replace `k` with `m * b`.
+  · rw [mul_assoc]
+    apply dvd_mul_right
+  . rw [mul_comm, mul_assoc]
+    apply dvd_mul_right
+
+
+-- You can use this to split into cases, depending on whether a proposition is true or false.
+-- To do this explicitly, you can use the law of the excluded middle:
+#check em
+
+example (P : Prop) : ¬¬P → P := by
+  intro h
+  rcases em P with hp | hnp
+  · assumption
+  . contradiction -- This tactic looks for a contradiction in the hypotheses, in this case `h` and `hnp`.
+
+-- You can also do this using the `by_cases` tactic.
+example (P : Prop) : ¬¬P → P := by
+  intro h
+  by_cases h' : P
```

Comment:
Maybe we should explain the syntax of the tactic as it's different from what seen before